specs/monitor_ceph_pool_utilization.adoc

```diff
@@ -0,0 +1,192 @@
+import rados
+import json
+
+def get_pool_utilization(r):
+    cmd = {"prefix":"df", "format":"json"}
+    ret, buf, errs = r.mon_command(json.dumps(cmd), b'', timeout=5)
+    result = json.loads(buf)
+    return result
+
+try:
+    r = rados.Rados(conffile="/etc/ceph/ceph.conf")
+    r.connect()
+
+    for i in range(1000):
+        print(get_pool_utilization(r))
+        time.sleep(1)
+except KeyboardInterrupt:
+    pass
+finally:
+    if r is not None:
+        r.shutdown()
+```
+
+* Option-2: The other option is to use execute command `ceph df` as a process
+and parse the output for data
+
+```
+import json
+import subprocess
+
+cluster = 'my_cluster'
+cmd = [
+    'ceph',
+    '--cluster',
+    cluster,
+    '-f',
+    'json-pretty',
+    'df'
+]
+process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+out, err = process.communicate()
+ret_code = process.poll()
+if ret_code:
+    print "command %s exited with non-zero status: %d" % (cmd, ret_code)
+    return
+
+df = json.loads(out)
+```
+
+=== Data model impact:
+
+None
+
+=== Impacted Modules:
+
+==== Tendrl API impact:
+
+* The performance monitoring application will periodically update the utilization
+data to a location in etcd and also the prformance monitoring application will
+expose APIs to access this data and the tendrl-api will proxy to it.
+
+==== Notifications/Monitoring impact:
+
+* Alerting module to handle the raising of appropriate alerts and notifications
+in case of threshold breaches of ceph pool utilization limits.
+
+==== Tendrl/common impact:
+
+None
+
+==== Tendrl/node_agent impact:
+
+None
+
+==== Sds integration impact:
+
+None
+
+=== Security impact:
+
+None
+
+=== Other end user impact:
+
+* Th trending graphs of ceph pool utilization data would help administrators to
+decide on bringing in additional storage capacity in system based on warning and
+critical threshold breaches depicted clearly in Tendrl UI
+
+=== Performance impact:
+
+* As confirmed by eph team running of `ceph df` command on mon nodes doesnt have
+any major impact as its not heavy and fetches details from in memory mon details
+
+=== Other deployer impact:
+
+None
+
+=== Developer impact:
+
+* API module owner(s) to take care of any API impacts
+
+* Alerting module owner(s) to take care handling of threshold breaches for ceph
+pool utilization
+
+== Implementation:
+
+* Add a collectd plugin which runs `ceph df` command at regular interval and
+pushes the data to time series DB
+
+* Out of two alternatives both are equally good, but in this scenario usage of
+option-2 would be more meaningful
+
+* Make sure interval for running this plugin is nothing less than 5 seconds as
+ceph mon use this intervals for fetching details from underlying cluster. So in
+ideal situation the interval for this plugin should be anything but more than 5
+seconds
+
+* Configure this plugin only if monitoring feature is enabled for Tendrl
+
+* Deploy the plugin only on the mon nodes
```

Comment:
The plugin needs to be configured only on the leader mon node(only 1 node) because if it it is running on multiple nodes of a cluster, all of them will raise threshold breach alerts(if any) for the same set of pools in cluster and hence there will be as many duplicate alerts as the number of mons in cluster(if deployed on all mons) which can be handled internally(if un-avoidable) but it might be inefficient.